File: src/dp/_516_longest_palindrome_subseq.rs
struct Solution;

impl Solution {
    pub fn longest_palindrome_subseq(s: String) -> i32 {
        let s = s.chars().collect::<Vec<_>>();
        let mut dp = vec![vec![0; s.len()]; s.len()];
        for i in (0..s.len()).rev() {
            dp[i][i] = 1;
            for j in i + 1..s.len() {
                if s[i] == s[j] {
                    dp[i][j] = dp[i + 1][j - 1] + 2;
                } else {
                    dp[i][j] = dp[i + 1][j].max(dp[i][j - 1]);
                }
            }
        }
        dp[0][s.len() - 1]
    }


    // todo 用状态压缩进行优化
    pub fn longest_palindrome_subseq1(s: String) -> i32 {
        0
    }
}

File: leetcode-go/516_longestPalindromeSubseq.go
package leetcode_go

func longestPalindromeSubseq(s string) int {
	dp := make([][]int, len(s))
	for i := 0; i < len(s); i++ {
		for j := 0; j < len(s); j++ {
			if dp[i] == nil {
				dp[i] = make([]int, len(s))
			}
			if i == j {
				dp[i][j] = 1
			}
		}
	}
	for i := len(s); i >= 0; i-- {
		for j := i + 1; j < len(s); j++ {
			if s[i] == s[j] {
				dp[i][j] = dp[i+1][j-1] + 2
			} else {
				dp[i][j] = max(dp[i+1][j], dp[i][j-1])
			}
		}
	}
	return dp[0][len(s)-1]
}
